File: Ponykart/Core/Options.cs
﻿using System;
using System.Collections.Generic;
using System.IO;
using Mogre;

namespace Ponykart.Core {
	public static class Options {
		private static IDictionary<string, string> dict;
		private static IDictionary<string, string> defaults;
		public static ModelDetailOption ModelDetail;
		public static ShadowDetailOption ShadowDetail;

		/// <summary>
		/// Creates the folder and file if they don't exist, and either prints some data to it (if it doesn't exist) or reads from it (if it does)
		/// </summary>
		public static void Initialise() {
			SetupDictionaries();
#if DEBUG
			string pkPath = Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData) + "\\Ponykart";

			// if a folder doesn't exist there, create it
			if (!Directory.Exists(pkPath))
				Directory.CreateDirectory(pkPath);

			string optionsPath = pkPath + "\\options.ini";
#else
			string optionsPath = "options.ini";
#endif

			// create it if the file doesn't exist, and write out some defaults
			if (!File.Exists(optionsPath)) {
				using (FileStream stream = File.Create(optionsPath)) {
					using (StreamWriter writer = new StreamWriter(stream)) {
						foreach (KeyValuePair<string, string> pair in defaults) {
							writer.WriteLine(pair.Key + "=" + pair.Value);
						}
					}
				}
				ModelDetail = ModelDetailOption.Medium;
			}
			// otherwise we just read from it
			else {
				ConfigFile cfile = new ConfigFile();
				cfile.Load(optionsPath, "=", true);

				ConfigFile.SectionIterator sectionIterator = cfile.GetSectionIterator();
				sectionIterator.MoveNext();
				foreach (KeyValuePair<string, string> pair in sectionIterator.Current) {
					dict[pair.Key] = pair.Value;
				}
				ModelDetail = (ModelDetailOption) Enum.Parse(typeof(ModelDetailOption), dict["ModelDetail"], true);
				ShadowDetail = (ShadowDetailOption) Enum.Parse(typeof(ShadowDetailOption), dict["ShadowDetail"], true);

				cfile.Dispose();
				sectionIterator.Dispose();
			}

#if DEBUG
			// since we sometimes add new options, we want to make sure the .ini file has all of them
			Save();
#endif
		}

		private static void SetupDictionaries() {
			// set up our dictionary with some default stuff in it
			defaults = new Dictionary<string, string>(){
				{"FSAA","0"},
				// Fastest or Accurate
				{"Floating-point mode","Fastest"},
				// Yes or No
				{"Full Screen","No"},
				// Yes or No
				{"VSync","Yes"},
				// 1, 2, 3, or 4
				{"VSync Interval","1"},
				{"Video Mode","1280 x 800 @ 32-bit colour"},
				// Yes or No
				{"sRGB Gamma Conversion","No"},
				// Yes or No
				{"Music","No"},
				{"Sounds","Yes"},
				{"Ribbons","Yes"},
				// Low or High
				{"ModelDetail","Medium"},
				{"ShadowDetail","Some"},
				{"ShadowDistance","40"},
				{"Twh","No"},
				{"Controller", "Keyboard"},
			};
			// copy it into the regular dictionary
			dict = new Dictionary<string, string>(defaults);
		}

		/// <summary>
		/// Writes out the current settings to our options file
		/// </summary>
		public static void Save() {
#if DEBUG
			string optionsPath = Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData) + "\\Ponykart\\options.ini";
#else
			string optionsPath = "options.ini";
#endif
			dict["ModelDetail"] = ModelDetail.ToString();
			dict["ShadowDetail"] = ShadowDetail.ToString();

			using (FileStream stream = File.Create(optionsPath)) {
				using (StreamWriter writer = new StreamWriter(stream)) {
					foreach (KeyValuePair<string, string> pair in dict) {
						writer.WriteLine(pair.Key + "=" + pair.Value);
					}
				}
			}
		}

		/// <summary>
		/// Gets an option.
		/// </summary>
		public static string Get(string keyName) {
			if (string.Equals(keyName, "ModelDetail", StringComparison.InvariantCultureIgnoreCase))
				throw new ArgumentException("Use the Options.ModelDetail enum instead of this method!", "keyName");
			else if (string.Equals(keyName, "ShadowDetail", StringComparison.InvariantCultureIgnoreCase))
				throw new ArgumentException("Use the Options.ShadowDetail enum instead of this method!", "keyName");

			return dict[keyName];
		}

		/// <summary>
		/// Gets an option as a boolean.
		/// </summary>
		public static bool GetBool(string keyName) {
			string value = dict[keyName];
			if (value == "Yes")
				return true;
			else if (value == "No")
				return false;
			else
				throw new ArgumentException("That key does not represent a boolean option!", "keyName");
		}

		/// <summary>
		/// Gets the dimensions of the render window
		/// </summary>
		/// <param name="height">The height of the window</param>
		/// <param name="width">The width of the window</param>
		public static void GetWindowDimensions(out uint width, out uint height) {
			string videoMode = dict["Video Mode"];
			string[] split = videoMode.Split(' ');

			width = uint.Parse(split[0]);
			height = uint.Parse(split[2]);
		}
	}
}

File: Ponykart++/Core/Options.cpp
#include <fstream>
#include "Core/Options.h"
#include "Ogre.h"

using namespace Ponykart;
using namespace Ponykart::Core;
using namespace Ogre;
using namespace std;

// Define the static members
unordered_map<string, string> Options::dict;
unordered_map<string, string> Options::defaults;
ModelDetailOption Options::ModelDetail;
ShadowDetailOption Options::ShadowDetail;

void Options::SetupDictionaries()
{
	defaults={
		{"FSAA","0"},
		{"Floating-point mode","Fastest"}, // Fastest or Accurate
		{"Full Screen","No"}, // Yes or No
		{"VSync","Yes"}, // Yes or No
		{"VSync Interval","1"}, // 1, 2, 3, or 4
		{"Video Mode","1024 x 768 @ 32-bit colour"},
		{"sRGB Gamma Conversion","No"}, // Yes or No
		{"Music","No"}, // Yes or No
		{"Sounds","Yes"},
		{"Ribbons","Yes"},
		{"ModelDetail","Medium"}, // Low, Medium or High
		{"ShadowDetail","Some"}, // None, Some or Many
		{"ShadowDistance","40"},
		{"Twh","No"},
		{"Controller", "Keyboard"},
	};
	dict = defaults; // copy it into the regular dictionary
}

void Options::Initialize()
{
	SetupDictionaries();

	constexpr char optionsPath[] = "media/config/ponykart.cfg";

	fstream file;
	file.open(optionsPath,ios::in);
	if (!file.is_open()) // create it if the file doesn't exist, and write out some defaults
	{
		file.open(optionsPath,ios::out);
		if (!file.is_open()) throw string("Cannot initialize media/config/ponykart.cfg");
		for (auto setting : defaults)
			file << setting.first << "=" << setting.second << endl;
		file.close();
		ModelDetail = ModelDetailOption::Medium;
		ShadowDetail = ShadowDetailOption::Some;
	}
	else // otherwise we just read from it
	{
		file.close(); // We're going to open it with cfile
		ConfigFile cfile;
		cfile.load(optionsPath, "=", true);
		auto sectionIterator = cfile.getSectionIterator();
		ConfigFile::SettingsMultiMap* settings=sectionIterator.getNext();
		for (auto curPair : *settings)
			dict[curPair.first] = curPair.second;
		// Convert the settings for ModelDetail and ShadowDetail (string) into enum values (int)
		string enumStr = dict["ModelDetail"];
		ModelDetail = enumStr=="High"?ModelDetailOption::High: (enumStr=="Low"?ModelDetailOption::Low: (ModelDetailOption::Medium));
		enumStr = dict["ShadowDetail"];
		ShadowDetail = enumStr=="Many"?ShadowDetailOption::Many: (enumStr=="None"?ShadowDetailOption::None: (ShadowDetailOption::Some));
	}

#ifdef DEBUG
	// since we sometimes add new options, we want to make sure the .ini file has all of them
	Save();
#endif
}

void Options::Save()
{
	constexpr char optionsPath[] = "media/config/ponykart.cfg";

	dict["ModelDetail"] = ModelDetail==ModelDetailOption::High?"High": (ModelDetail==ModelDetailOption::Low?"Low": ("Medium"));
	dict["ShadowDetail"] = ShadowDetail==ShadowDetailOption::Many?"Many": (ShadowDetail==ShadowDetailOption::None?"None": ("Some"));

	fstream file;
	file.open(optionsPath,ios::out);
	if (!file.is_open()) throw string("Cannot save media/config/ponykart.cfg");

	for (auto setting : dict)
		file << setting.first << "=" << setting.second << endl;
}

string Options::Get(const string keyName)
{
	// TODO: The string comparisons should be case-insensitive, like in the C# code
	if (keyName == "ModelDetail")
		throw new string("Options::getBool: Use the Options.ModelDetail enum instead of this method!");
	else if (keyName == "ShadowDetail")
		throw new string("Use the Options.ShadowDetail enum instead of this method!");

	return dict[keyName];
}
